history/cycleValidator.go

```diff
@@ -0,0 +1,261 @@
+// (except for the most recent session) have an offline event. It takes in excludeFunc, which is a function that
+// takes in a session ID and returns true if that session is still valid even if it does not have an offline event.
+func SessionOfflineValidator(excludeFunc func(id string) bool) CycleValidatorFunc {
+	return func(events []interpreter.Event) (bool, error) {
+		if len(events) == 0 {
+			return true, nil
+		}
+
+		offlineEvents := make(map[string]bool)
+		sessionIDs := make(map[string]bool)
+		for _, event := range events {
+			sessionID := event.SessionID
+			eventType, err := event.EventType()
+			if len(sessionID) == 0 || err != nil {
+				continue
+			}
+
+			sessionIDs[sessionID] = true
+
+			if eventType == interpreter.OfflineEventType {
+				offlineEvents[sessionID] = true
+			}
+
+		}
+
+		if valid, invalidIds := determineError(sessionIDs, offlineEvents, excludeFunc); valid {
+			return true, nil
+		} else {
+			return false, CycleValidationErr{
+				OriginalErr:   ErrMissingOfflineEvent,
+				InvalidFields: invalidIds,
+				ErrorTag:      validation.MissingOfflineEvent,
+			}
+		}
+	}
+}
+
+func determineError(sessionIDs map[string]bool, existingEvents map[string]bool, exclude func(id string) bool) (bool, []string) {
+	var missingEvents []string
+	for id := range sessionIDs {
+		if !existingEvents[id] && !exclude(id) {
+			missingEvents = append(missingEvents, id)
+		}
+	}
+
+	if len(missingEvents) == 0 {
+		return true, nil
+	}
+
+	return false, missingEvents
+}
+
+func determineMetadataValues(fields []string, event interpreter.Event) map[string]string {
+	values := make(map[string]string)
+	for _, field := range fields {
+		if val, ok := event.Metadata[field]; ok {
+			values[field] = val
+		} else {
+			values[field] = ""
+		}
+	}
+
+	return values
+}
+
+func validateMetadata(keys []string, events []interpreter.Event) []string {
+	if len(events) == 0 {
+		return nil
+	}
+
+	metadataVals := determineMetadataValues(keys, events[0])
+	incorrectFieldsMap := make(map[string]bool)
+	for _, event := range events {
+		incorrectFieldsMap = checkMetadataValues(metadataVals, incorrectFieldsMap, event)
+	}
+
+	if len(incorrectFieldsMap) == 0 {
+		return nil
+	}
+
+	fields := make([]string, 0, len(incorrectFieldsMap))
+	for key := range incorrectFieldsMap {
+		fields = append(fields, key)
+	}
+
+	return fields
+
+}
+
+func validateMetadataWithinCycle(keys []string, events []interpreter.Event) []string {
+	if len(events) == 0 {
+		return nil
+	}
+
+	metadataVals := make(map[int64]map[string]string)
+	incorrectFieldsMap := make(map[string]bool)
+	for _, event := range events {
+		boottime, err := event.BootTime()
+		if err != nil || boottime <= 0 {
+			continue
+		}
+
+		expectedVals, found := metadataVals[boottime]
+		if !found {
+			expectedVals = determineMetadataValues(keys, event)
+			metadataVals[boottime] = expectedVals
+			continue
```

Comment:
This was a little confusing, since I thought the `expectedVals` would be needed in the scope it was declared in, but we `continue` and it isn't used in that scope again.  My suggestion would be declare a different shorter variable name for keeping the lines split up, or just put them together, like below:
```suggestion
			metadataVals[boottime] = determineMetadataValues(keys, event)
			continue
```

Reply:
Comments in this function may also help.